src/analysis/calculus/inverse.lean

```diff
@@ -0,0 +1,582 @@
+variables {E : Type*} [normed_group E] [normed_space 𝕜 E]
+variables {F : Type*} [normed_group F] [normed_space 𝕜 F]
+variables {G : Type*} [normed_group G] [normed_space 𝕜 G]
+variables {G' : Type*} [normed_group G'] [normed_space 𝕜 G']
+
+open asymptotics filter metric set
+open continuous_linear_map (id)
+
+/-- We say that `f` approximates continuous linear map `f'` on `s` with constant `c`,
+if `∥f x - f y - f' (x - y)∥ ≤ c * ∥x - y∥` whenever `x, y ∈ s`.
+
+This predicate is defined to faciliate splitting of the inverse function theorem into small lemmas.
+Some of these lemmas can be useful, e.g., to prove that the inverse function is defined
+on a specific set. -/
+def approximates_linear_on (f : E → F) (f' : E →L[𝕜] F) (s : set E) (c : ℝ≥0) : Prop :=
+∀ (x ∈ s) (y ∈ s), ∥f x - f y - f' (x - y)∥ ≤ c * ∥x - y∥
+
+namespace approximates_linear_on
+
+variables [cs : complete_space E] {f : E → F}
+
+section
+
+variables {f' : E →L[𝕜] F} {s t : set E} {c c' : ℝ≥0}
+
+theorem mono_num (hc : c ≤ c') (hf : approximates_linear_on f f' s c) :
+  approximates_linear_on f f' s c' :=
+λ x hx y hy, le_trans (hf x hx y hy) (mul_le_mul_of_nonneg_right hc $ norm_nonneg _)
+
+theorem mono_set (hst : s ⊆ t) (hf : approximates_linear_on f f' t c) :
+  approximates_linear_on f f' s c :=
+λ x hx y hy, hf x (hst hx) y (hst hy)
+
+lemma lipschitz_sub (hf : approximates_linear_on f f' s c) :
+  lipschitz_with c (λ x : s, f x - f' x) :=
+begin
+  refine lipschitz_with.of_dist_le_mul (λ x y, _),
+  rw [dist_eq_norm, subtype.dist_eq, dist_eq_norm],
+  convert hf x x.2 y y.2 using 2,
+  rw [f'.map_sub], abel
+end
+
+protected lemma lipschitz (hf : approximates_linear_on f f' s c) :
+  lipschitz_with (nnnorm f' + c) (s.restrict f) :=
+by simpa only [restrict_apply, add_sub_cancel'_right]
+  using (f'.lipschitz.restrict s).add hf.lipschitz_sub
+
+protected lemma continuous (hf : approximates_linear_on f f' s c) :
+  continuous (s.restrict f) :=
+hf.lipschitz.continuous
+
+protected lemma continuous_on (hf : approximates_linear_on f f' s c) :
+  continuous_on f s :=
+continuous_on_iff_continuous_restrict.2 hf.continuous
+
+end
+
+variables {f' : E ≃L[𝕜] F} {s : set E} {c : ℝ≥0}
+  (hf : approximates_linear_on f (f' : E →L[𝕜] F) s c)
+
+include hf
+
+local notation `N` := nnnorm (f'.symm : F →L[𝕜] E)
+
+variable (hc : subsingleton E ∨ c < N⁻¹)
+
+include hc
+
+protected lemma antilipschitz : antilipschitz_with (N⁻¹ - c)⁻¹ (s.restrict f) :=
+begin
+  cases hc with hE hc,
+  { haveI : subsingleton s := ⟨λ x y, subtype.eq $ @subsingleton.elim _ hE _ _⟩,
+    exact antilipschitz_with.of_subsingleton },
+  convert (f'.antilipschitz.restrict s).add_lipschitz_with hf.lipschitz_sub hc,
+  simp [restrict]
+end
+
+protected lemma injective : injective (s.restrict f) :=
+(hf.antilipschitz hc).injective
+
+protected lemma inj_on : inj_on f s :=
+inj_on_iff_injective.2 $ hf.injective hc
+
+/-- A map approximating a linear equivalence defines a local equivalence. Should not
```

Comment:
```suggestion
/-- A map approximating a linear equivalence on a set defines a local equivalence on this set. Should not
```
(It was not obvious from the docstring that one did not have to restrict furhter the source).